### PR TITLE
Hubble: improve security by adding an option to redact API key in Kafka requests (L7)

### PR DESCRIPTION
**Pull request by ioandr**
Add option to redact API key in Kafka requests.

### TODOS

- [X] Accept option to redact Kafka API key
- [X] Add business logic to L7 parser
- [X] Add unit tests
- [X] Update Helm charts
- [X] Update docs 

Refs: https://github.com/cilium/cilium/issues/23887

Depends on: https://github.com/cilium/cilium/pull/25746

### CODE REVIEW

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks for the PR @ioandr, patch LGTM.

I would like to see a new test ensuring that `KafkaRedactAPIKey` is honored. Also although a breaking change I think it should be `true` by default cc @rolinh.

File: pkg/hubble/parser/seven/kafka.go
Comment (rolinh) [inline pkg/hubble/parser/seven/kafka.go]: Unfortunately, we can only allow read access to the LogRecord structure and thus we should only change the value that is assigned to the flow's `Kafka.ApiKey` (see [here](https://github.com/cilium/cilium/blob/a899cf05a4b220568f142f6147951edea23ce10d/pkg/hubble/parser/seven/parser.go#L77-L80) for more context).

**Review by rolinh — CHANGES_REQUESTED** [review]
Thanks for the patch! Unfortunately, we can't mutate LogRecord structures so the patch needs to be updated (see comment below).

> Also although a breaking change I think it should be true by default cc @rolinh.

Mmh, I'm not sure about that. If L7 visibility was an opt-out feature, then I'd clearly say yes, Hubble should redact Kafka's API keys by default. However, L7 visibility is opt-in and its (security) implications should be well understood when enabling it (note: we should probably update the documentation to make that clear). My point is that even if we try our best to redact sensitive data, there's no guarantee that Hubble will always correctly redact all sensitive data from L7 payloads. Therefore, L7 flows should be treated as sensitive data overall. I also think that if we redact (known) sensitive data by default, then that means we should do it overall which would imply that we also redact all HTTP headers by default and I'm not sure we would want that.

File: pkg/hubble/parser/seven/kafka.go
Comment (ioandr) [reply pkg/hubble/parser/seven/kafka.go]: Hi @rolinh, thanks for your feedback and for guiding the overall implementation.

>  [...] L7 visibility is opt-in and its (security) implications should be well understood when enabling it (note: we should probably update the documentation to make that clear)

Totally agree, it definitely makes sense to include this in the docs. Is the following guide the right place to add this piece of information?

https://docs.cilium.io/en/stable/observability/visibility/

Maybe we could add a new subsection named "Security" above "Troubleshooting" and "Limitations". WDYT?

> I also think that if we redact (known) sensitive data by default, then [...] and I'm not sure we would want that.

Based on this approach I understand that we should **not** redact sensitive info present in L7 flows by default. That is, we should set the default value of the newly introduced `KafkaRedactAPIKey` option to `false`.

**Comment by kaworu** [comment]
@rolinh generally agree with all that you wrote

> I also think that if we redact (known) sensitive data by default, then that means we should do it overall which would imply that we also redact all HTTP headers by default and I'm not sure we would want that.

HTTP headers may or may not be sensitive data, while I think we can agree basic auth password and Kafka API key always are. We're already redacting the former unconditionally, so in my opinion doing it by default for the latter would be consistent. Not feeling too strongly about this though.

File: pkg/hubble/parser/seven/kafka.go
Comment (kaworu) [reply pkg/hubble/parser/seven/kafka.go]: A "Security Implications" section at https://docs.cilium.io/en/stable/observability/visibility/ sounds good to me, please document on the way that we're redacting HTTP basic auth password :pray:

**Review by kaworu — APPROVED** [review]
Thanks @ioandr for the quick update! The patch LGTM besides the missing doc update :rocket:

File: pkg/hubble/parser/seven/kafka.go
Comment (ChrsMark) [reply pkg/hubble/parser/seven/kafka.go]: Kind of related: isn't the mutation of the HTTP LogRecord at https://github.com/cilium/cilium/blob/e727175569f99c988d6474b9ca457fd082ec51dd/pkg/hubble/parser/seven/http.go#L33 
violating this "read-only" rule? Came across this while at https://github.com/cilium/cilium/pull/25746/files#diff-1df644554f9d2598342e7c4d8f9f3ae5af5fd29b52ed0dc88e01ea986f218a36R37 :)

**Comment by ioandr** [comment]
> Thanks @ioandr for the quick update! The patch LGTM besides the missing doc

Sure, I can prepare a short doc as discussed above.

However the following is not 100% clear to me:

How are we planning to expose this option to end-users? That is, do we want to support allowing end-users to control this option to via the `cilium` CLI? For instance:

```
$ cilium config set hubble-l7-parser-kafka-redact-api-key true
```

Looking at the codebase, this requires the following:

1. Extend Cilium https://github.com/cilium/cilium/blob/main/pkg/option/config.go with an option to control whether Hubble's L7 parser redacts Kafka API keys - we already declare `Hubble*` options there
2. Set a sane default value for the newly added Hubble L7 option under https://github.com/cilium/cilium/blob/main/pkg/defaults/defaults.go
3. Pass context to the L7 payload parser upon creation: https://github.com/cilium/cilium/blob/main/daemon/cmd/hubble.go#L141
4. Possibly update Cilium's Helm chart as needed: https://docs.cilium.io/en/stable/helm-reference/

This is what @ChrsMark has done in his PR that tackles the redaction of HTTP query params, so maybe we want a uniform approach in both. PTAL: https://github.com/cilium/cilium/pull/25746/files

CC @kaworu @rolinh

File: Documentation/observability/visibility.rst
Comment (kaworu) [inline Documentation/observability/visibility.rst]: Without a set plan for theses feature (i.e. deadline or event release target) I would prefer to avoid mentioning them in the documentation.

```suggestion
```

File: Documentation/observability/visibility.rst
Comment (kaworu) [inline Documentation/observability/visibility.rst]: IIUC this is part of https://github.com/cilium/cilium/pull/25746 so let's wait for it to be merged before merging this PR.

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks @ioandr and @ChrsMark!

I think we should merge https://github.com/cilium/cilium/pull/25746 first, as it touches on HTTP for which we already redact the basic auth password, so it'll be easier to bring consistency starting from there (e.g. avoid mutating accesslog https://github.com/cilium/cilium/pull/25844#discussion_r1220380104).

About enabling/disabling, I think one cilium cli flag should be added instead of one per sensitive data redaction, otherwise we'll bloat the cli flag list. We can take inspiration of the current [Hubble metric configuration](https://github.com/cilium/cilium/blob/3f78ab717d855f4a376bf34606430280a902b9c8/install/kubernetes/cilium/templates/cilium-configmap.yaml#L813-L822) (see also [here](https://github.com/cilium/cilium/blob/3f78ab717d855f4a376bf34606430280a902b9c8/daemon/cmd/daemon_main.go#L935-L936)). Something along the line of:

```
--hubble-redact=kafa-api-key
--hubble-redact=http-basic-auth-password
--hubble-redact=http-url-query
```

What do you think? cc @rolinh

File: Documentation/observability/visibility.rst
Comment (ioandr) [reply Documentation/observability/visibility.rst]: @kaworu this makes sense, we will remove those from the doc.

**Comment by ioandr** [comment]
> Thanks @ioandr and @ChrsMark!
> 
> I think we should merge #25746 first, as it touches on HTTP for which we already redact the basic auth password, so it'll be easier to bring consistency starting from there (e.g. avoid mutating accesslog [#25844 (comment)](https://github.com/cilium/cilium/pull/25844#discussion_r1220380104)).
> 
> About enabling/disabling, I think one cilium cli flag should be added instead of one per sensitive data redaction, otherwise we'll bloat the cli flag list. We can take inspiration of the current [Hubble metric configuration](https://github.com/cilium/cilium/blob/3f78ab717d855f4a376bf34606430280a902b9c8/install/kubernetes/cilium/templates/cilium-configmap.yaml#L813-L822) (see also [here](https://github.com/cilium/cilium/blob/3f78ab717d855f4a376bf34606430280a902b9c8/daemon/cmd/daemon_main.go#L935-L936)). Something along the line of:
> 
> ```
> --hubble-redact=kafa-api-key
> --hubble-redact=http-basic-auth-password
> --hubble-redact=http-url-query
> ```
> 
> What do you think? cc @rolinh

Hi @kaworu @rolinh 

we did some work with @ChrsMark based on your feedback, PTAL https://github.com/cilium/cilium/pull/25746.

Let's move the review/discussion there and rebase the current (Kafka) PR once #25746 is merged.

File: Documentation/observability/visibility.rst
Comment (zacharysarah) [inline Documentation/observability/visibility.rst]: Edits for clarity, no Latin abbreviations
```suggestion
Monitoring Layer 7 traffic involves security considerations for handling potentially sensitive information, such as usernames, passwords, query parameters, API keys, and others.
```

File: Documentation/observability/visibility.rst
Comment (zacharysarah) [inline Documentation/observability/visibility.rst]: Indent, omit bold
```suggestion
    By default, Hubble does not protect potentially sensitive information
    present in `Layer 7 Hubble Flows <https://github.com/cilium/cilium/tree/master/api/v1/flow#flow-Layer7>`_.
```

File: Documentation/observability/visibility.rst
Comment (zacharysarah) [inline Documentation/observability/visibility.rst]: Omit bold
```suggestion
* HTTP: redact URL query parameters
* Kafka: redact API key
```

File: Documentation/observability/visibility.rst
Comment (zacharysarah) [inline Documentation/observability/visibility.rst]: Specific is better than general, comma nit
```suggestion
For more information on configuring Cilium, see :ref:`Cilium Configuration <configuration>`.
```

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl:1015]: Good use of present tense, active voice, and addressing the user as "you" ✨

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@ioandr Good start! Some small changes, otherwise LGTM for docs

**Comment by ioandr** [comment]
Hi @zacharysarah thanks for your comments, they are spot on!

Note that in this PR we are adding to the "Security Implications" section of the `visibility.rst` document that https://github.com/cilium/cilium/pull/25746 introduces.

IMO it makes sense to apply suggested changes there, rebase the current PR on top of it and finally stick to your guidelines for adding Kafka-specific content.

CC @ChrsMark

**Comment by ioandr** [comment]
Since https://github.com/cilium/cilium/pull/25746 got merged into `main` I rebased the current PR and resolved conflicts so that we follows the same conventions and guidelines (naming, documentation, CLI option) for both HTTP and Kafka.

LMK there are any actions needed from our side for the review.

CC @rolinh @kaworu @zacharysarah @qmonnet

**Review by qmonnet — APPROVED** [review]

**Review by rolinh — APPROVED** [review]
LGTM, thanks!

**Comment by ChrsMark** [comment]
Hey @pippolo84 @nathanjsweet @asauber @aditighag any chances having a look into this one? It should be a simple one 🙏🏽. Despite being labeled with `dont-merge/wait-until-release` it would be great if can have it reviewed and ready for merge :).

**Review by zacharysarah — APPROVED** [review]
LGTM

**Review by pippolo84 — APPROVED** [review]
Thanks! :100:

**Review by jrajahalme — APPROVED** [review]
LGTM

**Review by kaworu — APPROVED** [review]
Thanks for the update @ioandr patch LGMT!

> @kaworu Looks like your concern was addressed?

Yes, although I still find it inconsistent to redact the HTTP password unconditionally while the Kafka API key redaction is opt-in, but that's arguably out of the scope of this PR.

**Comment by ldelossa** [comment]
@ioandr, Can you please rebase this PR on `main` and push? Should fix some issues with the tests.

**Comment by ioandr** [comment]
@ldelossa the only test that is failing is the mergeability one - I suppose due to the `dont-merge/wait-until-release` label that prevents merge.

LMK if you need any further action from my side.

**Review by asauber — APPROVED** [review]
LGTM, my main feedback here would be that we should list the "allowed" redacted values more clearly in the docs; make it clear that right now there are exactly two options and what they are.

**Comment by ioandr** [comment]
> LGTM, my main feedback here would be that we should list the "allowed" redacted values more clearly in the docs; make it clear that right now there are exactly two options and what they are.

Hi @asauber, thank you for your feedback on this. How about we rephrase the doc as follows:

```
To harden security, Cilium provides the ``--hubble-redact`` option which
accepts a comma-separated list of values that control how Hubble handles
sensitive information present in Layer 7 flows. More specifically, it offers
the following features and values for supported Layer 7 protocols:

* For HTTP: redacting URL query (GET) parameters (``--hubble-redact=http-url-query``)
* For Kafka: redacting API key (``--hubble-redact=kafka-api-key``)
```

**Review by nathanjsweet — APPROVED** [review]

**Comment by asauber** [comment]
@ioandr That phrasing looks good to me

**Comment by ioandr** [comment]
@ldelossa I see that you removed the `dont-merge/wait-until-release` label so I suppose there is no blocker towards merging this.

I see that [one test](https://github.com/cilium/cilium/actions/runs/5413874539/jobs/9840682841) is failing with:

```
connectivity test failed: timeout reached waiting lookup for localhost from pod cilium-test/client-6965d549d5-mjxk9 to server on pod cilium-test/echo-other-node-545c9b778b-m6d6t to succeed (last error: context deadline exceeded)
Error: Process completed with exit code 1.
```

I believe this is something temporary/flaky, that is, the test failure is not caused by the changes of this PR. Maybe we can re-trigger tests?

**Pull request merged**
